src/views/ProblemsDiagnosticCollection.ts

```diff
@@ -49,90 +52,211 @@ export class IssueDiagnostic extends vscode.Diagnostic {
+      : result.rule?.id
+  }
+}
+
 export class ProblemsDiagnosticCollection implements vscode.Disposable {
   private _collection: vscode.DiagnosticCollection = vscode.languages.createDiagnosticCollection('codacy')
 
-  private _currentIssues: PullRequestIssue[] | BranchIssue[] = []
+  private _currentApiIssues: { [key in string]: PullRequestIssue[] | BranchIssue[] } = {}
+  private _currentCliIssues: { [key in string]: ProcessedSarifResult[] } = {}
+
+  private _isAnalysisRunning: boolean = false
+  private _analysisDebounceTimeout: NodeJS.Timeout | undefined
 
   constructor(private readonly _repositoryManager: RepositoryManager) {
+    // load all API issues when the pull request is updated
     _repositoryManager.onDidUpdatePullRequest((pr) => {
       const newIssues = pr?.issues.filter((issue) => issue.deltaType === 'Added') || []
-      if (newIssues.length > 0) {
-        this.load(newIssues)
-      } else {
-        this.clear()
-      }
+      this.loadAPIIssues(newIssues)
     })
 
+    // load all API issues when the branch is updated
     _repositoryManager.branchIssues.onDidUpdateBranchIssues((issues) => {
-      if (issues.length > 0) {
-        this.load(issues)
-      } else {
-        this.clear()
-      }
+      this.loadAPIIssues(issues)
     })
 
-    GitProvider.instance?.onDidChangeTextDocument((e) => {
-      this.updatePositions(e.document)
+    GitProvider.instance?.onDidChangeTextDocument(async (e) => {
+      // avoid if the ddocument is a .git file
+      if (e.document.uri.fsPath.endsWith('.git')) return
+
+      // update positions of remote issues in the document
+      this.updateApiISsuesPositions(e.document)
+
+      // run local analysis for available tools
+      await this.runAnalysisAndUpdateDiagnostics(e.document)
     })
   }
 
-  public load(issues: PullRequestIssue[] | BranchIssue[]) {
-    // clear previous diagnostics
+  private updateDiagnostics() {
     this._collection.clear()
-    this._currentIssues = []
+    const filesWithApiIssues = Object.keys(this._currentApiIssues)
+    const filesWithCliIssues = Object.keys(this._currentCliIssues)
+    const allFiles = [...new Set([...filesWithApiIssues, ...filesWithCliIssues])]
+    allFiles.forEach((file) => {
+      const document = vscode.workspace.textDocuments.find((doc) => doc.uri.fsPath === file)
+      if (document) {
+        this.updateDocumentDiagnostics(document)
+      }
+    })
+  }
 
-    // group issues by file using lodash
-    const issuesByFile = groupBy(issues, (issue) => issue.commitIssue.filePath)
-    const baseUri = this._repositoryManager.rootUri?.path
+  private updateDocumentDiagnostics(document: vscode.TextDocument) {
+    this._collection.delete(document.uri)
+
+    // get API issues for the current document
+    const documentApiIssues = this._currentApiIssues[document.uri.fsPath] || []
+    const documentCliIssues = this._currentCliIssues[document.uri.fsPath] || []
+
+    // remove API issues found in the CLI issues
```

Comment:
so basically we are prioritizing CLI issues, correct?

Reply:
yup, those should be more up to date